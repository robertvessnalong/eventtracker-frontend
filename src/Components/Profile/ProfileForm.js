import { useContext, useState, useEffect } from 'react';
import UserContext from '../../Context/UserContext';
import { Form, Button, Alert, Row, Col, Spinner } from 'react-bootstrap';
import FlightHubApi from '../../API/api';

const ProfileForm = () => {
  const { user, setUser } = useContext(UserContext);
  const formInitalState = {
    firstName: user.firstName,
    lastName: user.lastName,
    email: user.email,
    password: '',
  };
  const [formData, setFormData] = useState(formInitalState);
  const [error, setError] = useState(false);
  const [loading, setLoading] = useState(true);
  const [errorMsg, setErrorMsg] = useState([]);
  const [success, setSuccess] = useState(false);
  const [successMsg, setSuccessMsg] = useState([]);

  useEffect(() => {
    if (user.firstName) {
      setLoading(false);
      setFormData(formInitalState);
    }

    if (error || success) {
      const errOrSuccess = error ? 'error' : 'success';
      let timerFunc = setTimeout(() => {
        if (errOrSuccess === 'error') {
          setError(false);
        } else {
          setSuccess(false);
        }
      }, 10000);
      return () => clearTimeout(timerFunc);
    }
  }, [error, success, user.firstName]);

  const handleChange = (e) => {
    const { name, value } = e.target;
    setFormData((data) => ({
      ...data,
      [name]: value,
    }));
  };

  if (loading) {
    return (
      <Row>
        <Col style={{ textAlign: 'center' }}>
          <Spinner animation='border' role='status'>
            <span className='visually-hidden'>Loading...</span>
          </Spinner>
        </Col>
      </Row>
    );
  }

  const handleSubmit = async (e) => {
    e.preventDefault();
    const res = await FlightHubApi.updateUser(user.uuid, formData);
    if (res.error) {
      setError(true);
      setErrorMsg(res.error);
    } else {
      const { firstName, lastName, email } = res.user;
      setUser(res.user);
      setSuccess(true);
      setSuccessMsg(res.success);
      setFormData({
        firstName,
        lastName,
        email,
        password: '',
      });
      setErrorMsg('');
      setError(false);
    }
  };

  return (
    <>
      <Form onSubmit={handleSubmit}>
        {error && <Alert variant='danger'>{errorMsg}</Alert>}
        {success && <Alert variant='success'>{successMsg}</Alert>}
        <Form.Group className='mb-3' controlId='firstName'>
          <Form.Label>First Name</Form.Label>
          <Form.Control
            type='text'
            placeholder='First Name'
            name='firstName'
            value={formData.firstName}
            onChange={handleChange}
          />
        </Form.Group>
        <Form.Group className='mb-3' controlId='lastName'>
          <Form.Label>Last Name</Form.Label>
          <Form.Control
            type='text'
            placeholder='Last Name'
            name='lastName'
            value={formData.lastName}
            onChange={handleChange}
          />
        </Form.Group>
        <Form.Group className='mb-3' controlId='formBasicEmail'>
          <Form.Label>Email address</Form.Label>
          <Form.Control
            type='email'
            placeholder='Enter email'
            name='email'
            value={formData.email}
            onChange={handleChange}
          />
        </Form.Group>
        <Form.Group className='mb-3' controlId='formBasicPassword'>
          <Form.Label>Password</Form.Label>
          <Form.Control
            type='password'
            placeholder='Password'
            name='password'
            value={formData.password}
            onChange={handleChange}
          />
          <Form.Text className='text-muted'>Enter Your Password</Form.Text>
        </Form.Group>
        <Button className='blue-btn' type='submit'>
          Submit
        </Button>
      </Form>
    </>
  );
};

export default ProfileForm;
