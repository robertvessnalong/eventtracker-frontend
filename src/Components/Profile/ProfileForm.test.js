import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import ProfileForm from './ProfileForm';
import TestUser from '../../Testing/TestUser';
import UserContext from '../../Context/UserContext';
import { BrowserRouter } from 'react-router-dom';

describe('ProfileForm Component', () => {
  // Smoke Test
  test('renders without crashing', async () => {
    render(
      <BrowserRouter>
        <UserContext.Provider value={{ user: TestUser }}>
          <ProfileForm />
        </UserContext.Provider>
      </BrowserRouter>
    );

    const label = screen.getByPlaceholderText(/First Name/i);
    expect(label).toBeInTheDocument();
  });

  test('update', () => {
    render(
      <UserContext.Provider value={{ user: TestUser }}>
        <ProfileForm />
      </UserContext.Provider>
    );
    fireEvent.change(screen.getByPlaceholderText('First Name'), {
      target: { value: 'Test' },
    });
    fireEvent.change(screen.getByPlaceholderText('Last Name'), {
      target: { value: 'Five' },
    });
    fireEvent.change(screen.getByPlaceholderText('Enter email'), {
      target: { value: 'testfive@gmail.com' },
    });
    fireEvent.change(screen.getByPlaceholderText('Password'), {
      target: { value: 'password' },
    });
    fireEvent.click(screen.getByText('Submit'));
  });

  // Snapshot
  test('matches snapshot', async () => {
    const { asFragment } = render(
      <UserContext.Provider value={{ user: TestUser }}>
        <ProfileForm />
      </UserContext.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
