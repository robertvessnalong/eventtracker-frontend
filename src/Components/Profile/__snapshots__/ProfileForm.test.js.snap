// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfileForm Component matches snapshot 1`] = `
<DocumentFragment>
  <form
    class=""
  >
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="firstName"
      >
        First Name
      </label>
      <input
        class="form-control"
        id="firstName"
        name="firstName"
        placeholder="First Name"
        type="text"
        value="Robert"
      />
    </div>
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="lastName"
      >
        Last Name
      </label>
      <input
        class="form-control"
        id="lastName"
        name="lastName"
        placeholder="Last Name"
        type="text"
        value="Long"
      />
    </div>
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="formBasicEmail"
      >
        Email address
      </label>
      <input
        class="form-control"
        id="formBasicEmail"
        name="email"
        placeholder="Enter email"
        type="email"
        value="robert@gmail.com"
      />
    </div>
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="formBasicPassword"
      >
        Password
      </label>
      <input
        class="form-control"
        id="formBasicPassword"
        name="password"
        placeholder="Password"
        type="password"
        value=""
      />
      <small
        class="text-muted form-text"
      >
        Enter Your Password
      </small>
    </div>
    <button
      class="blue-btn btn btn-primary"
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;
